components/scheduler/executors.go

```diff
@@ -23,41 +23,456 @@
 	"go.temporal.io/server/service/history/hsm"
+	"go.temporal.io/server/service/worker/scheduler"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/types/known/timestamppb"
 )
 
-func RegisterExecutor(
-	registry *hsm.Registry,
-	executorOptions TaskExecutorOptions,
-	config *Config,
-) error {
-	activeExec := taskExecutor{options: executorOptions, config: config}
-	return hsm.RegisterTimerExecutor(
-		registry,
-		activeExec.executeScheduleTask,
-	)
-}
-
 type (
 	TaskExecutorOptions struct {
+		fx.In
+		MetricsHandler    metrics.Handler
+		Logger            log.Logger
+		SpecBuilder       *scheduler.SpecBuilder
+		FrontendClient    workflowservice.WorkflowServiceClient
+		HistoryClient     resource.HistoryClient
+		NamespaceRegistry namespace.Registry
+		Config            *Config
 	}
 
 	taskExecutor struct {
-		options TaskExecutorOptions
-		config  *Config
+		TaskExecutorOptions
 	}
 )
 
-func (e taskExecutor) executeScheduleTask(
+var ErrSchedulerConflict = errors.New("concurrent scheduler task execution detected, unable to update scheduler state")
+
+func RegisterExecutor(
+	registry *hsm.Registry,
+	executorOptions TaskExecutorOptions,
+) error {
+	exec := taskExecutor{executorOptions}
+	if err := hsm.RegisterTimerExecutor(
+		registry,
+		exec.executeSchedulerWaitTask); err != nil {
+		return err
+	}
+	return hsm.RegisterImmediateExecutor(registry, exec.executeSchedulerRunTask)
+}
+
+func (e taskExecutor) executeSchedulerWaitTask(
 	env hsm.Environment,
 	node *hsm.Node,
-	task ScheduleTask,
+	task SchedulerWaitTask,
 ) error {
 	if err := node.CheckRunning(); err != nil {
 		return err
 	}
-	// TODO(Tianyu): Perform scheduler logic before scheduling self again
-	return hsm.MachineTransition(node, func(scheduler Scheduler) (hsm.TransitionOutput, error) {
+	return hsm.MachineTransition(node, func(scheduler *Scheduler) (hsm.TransitionOutput, error) {
 		return TransitionSchedulerActivate.Apply(scheduler, EventSchedulerActivate{})
 	})
 }
+
+func (e taskExecutor) executeSchedulerRunTask(
+	ctx context.Context,
+	env hsm.Environment,
+	ref hsm.Ref,
+	task SchedulerActivateTask,
+) error {
+	ns, err := e.NamespaceRegistry.GetNamespaceByID(namespace.ID(ref.WorkflowKey.NamespaceID))
+	if err != nil {
+		return fmt.Errorf("failed to get namespace by ID: %w", err)
+	}
+	ctx, cancel := context.WithTimeout(ctx, e.Config.ExecutionTimeout(ns.Name().String()))
+	defer cancel()
+
+	var s Scheduler
+	var prevArgs *schedspb.HsmSchedulerState
+	err = env.Access(ctx, ref, hsm.AccessRead, func(node *hsm.Node) error {
+		if err := node.CheckRunning(); err != nil {
+			return err
+		}
+		prevS, err := hsm.MachineData[*Scheduler](node)
+		prevArgs = prevS.HsmSchedulerState
+		if err != nil {
+			return err
+		}
+		// Ensure that this is copy, so later (stateful) processing does not update state machine without protection
+		s = Scheduler{
+			HsmSchedulerState: common.CloneProto(prevS.HsmSchedulerState),
```

Comment:
I would recommend instead extracting just the fields you need to execute the tasks but this is fine for now.